📖  Document clusterctl version support (#3449)
kubernetes-sigs/cluster-api: docs/book/src/reference/versions.md

The Kubeadm Control Plane requires the Kubeadm Bootstrap Provider.

#### clusterctl

It is strongly recommended using a version of [clusterctl](../clusterctl/overview.md) that exactly matches the version
of the Cluster API core provider; in case of upgrades, clusterctl should be upgraded first and used to upgrade

> all the other components.

Reviewer: Not sure what we're trying to communicate here, do we want users to upgrade clusterctl to the latest available version all the time?

Author: Yes, IMO this is the safest approach: upgrade clusterctl first, then use the latest clusterctl to upgrade other components

Reviewer: ```suggestion
It is strongly recommended using the latest version of [clusterctl](../clusterctl/overview.md) before attempting to upgrade your clusters.
```

How about something like this? I'd also open an issue if we don't have one already to have ways in clusterctl to check for its latest version

Author: Thinking back to this might be I have misunderstood the original question because the consideration does not apply to upgrades only .

I have in mind some recent use cases of uses trying to init a cluster with version 0.3.7 but using an older version of clusterctl, and failing due to the changes for envsubst; I do expect might happen in the near future as well, e.g. for the changes on cert-manager, or for the changes on the move logic.

This is why IMO the version of clusterctl should always be >= to the version of CAPI installed in the cluster (and as corollary, clusterctl should be updated before all the other components)

Reviewer 2: Yes. Ideally, the user should always update to the latest version of clusterctl before doing any operations in order to avoid unintended side effects. 
We had a discussion regarding clusterctl doing a version check ([see slack thread](https://kubernetes.slack.com/archives/C8TSNPY4T/p1594765826104200?thread_ts=1594765139.103900&cid=C8TSNPY4T)). 

I can create an issue regarding clusterctl doing a version check.

Reviewer 2: I created this issue regarding above: https://github.com/kubernetes-sigs/cluster-api/issues/3466

Author: I have aligned the doc to the issue above, please PTAL

<aside class="note warning">

<h1>Warning!</h1>

clusterctl was re-designed in the v1alpha3/v0.3 release series.

A clusterctl version in the v1alpha3/v0.3 is expected to manage Cluster API providers the current and all the previous 
patch releases in the same series; however, considering the project is still in alpha, we are not providing any guarantee

> for using an older version of clusterctl to install the newest patch releases. 

Reviewer 3: I wonder if we need to mention the v1alpha3 in this statement here, we mention it on the line above and it's likely to be true for v1alpha4, v1alpha5 etc as well right? 

Perhaps:

```
Each clusterctl version is expected to manage Cluster API providers with a matching release as well as all previous patch releases in the same series (eg. v0.3.5 is compatible with v0.3.0 through v0.3.5); ...
```
I think this is what you are trying to say here? I'm a bit confused though, as the provider implementations have their own versioning schemes don't they? Eg Azure is on v0.4.x now even though that's compatible with cluster-api v0.3.x?

Author: You got this almost right, the trick here is that the release of clusterctl is linked mostly to the release of CAPI (all the other provider are linked indirectly to CAPI via the contract)

Reviewer 3: Oh right, in which case, should we change the word `provider`? That makes me think of the providers, perhaps `components` or something is better to mean the core CAPI controllers?

Author: I have removed this note because now this is implicit in the previous statement
